From b748b1e753d77920b42e6dd94c17917f29bd8db5 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sun, 28 Dec 2025 22:33:46 -0500
Subject: [PATCH 471/473] winewayland: Refactor setting window icons.

Remove the send_message stuff to get the icon. It causes hangs for absolutely 0 reason and this code has been replaced upstream anyways...
---
 dlls/winewayland.drv/window.c | 96 ++++++++++++++---------------------
 1 file changed, 39 insertions(+), 57 deletions(-)

diff --git a/dlls/winewayland.drv/window.c b/dlls/winewayland.drv/window.c
index dcc9189248f..3e9eec72199 100644
--- a/dlls/winewayland.drv/window.c
+++ b/dlls/winewayland.drv/window.c
@@ -258,11 +258,33 @@ static BOOL wayland_win_data_create_wayland_surface(struct wayland_win_data *dat
     return TRUE;
 }
 
+static HICON get_icon_info(HICON icon, ICONINFO *ii)
+{
+    return icon && NtUserGetIconInfo(icon, ii, NULL, NULL, NULL, 0) ? icon : NULL;
+}
+
+static HICON get_window_icon(HWND hwnd, UINT type, HICON icon, ICONINFO *ret)
+{
+    icon = get_icon_info(icon, ret);
+    if (!icon)
+    {
+        UINT offset = (type == ICON_BIG) ? GCLP_HICON : GCLP_HICONSM;
+        icon = get_icon_info((HICON)NtUserGetClassLongPtrW(hwnd, offset), ret);
+        if (!icon && type == ICON_BIG)
+        {
+            icon = LoadImageW(0, (const WCHAR *)IDI_WINLOGO, IMAGE_ICON, 0, 0,
+                              LR_SHARED | LR_DEFAULTSIZE);
+            icon = get_icon_info(icon, ret);
+        }
+    }
+    return icon;
+}
+
 static void wayland_surface_update_state_toplevel(struct wayland_surface *surface)
 {
     BOOL processing_config = surface->processing.serial &&
                              !surface->processing.processed;
-    RECT rect = surface->window.rect;
+    const RECT *rect = &surface->window.rect;
 
     TRACE("hwnd=%p window_state=%#x %s->state=%#x\n",
           surface->hwnd, surface->window.state,
@@ -297,7 +319,7 @@ static void wayland_surface_update_state_toplevel(struct wayland_surface *surfac
         {
             struct wl_output *output;
             pthread_mutex_lock(&process_wayland.output_mutex);
-            output = wayland_get_best_output_for_rect(&surface->window.rect);
+            output = wayland_get_best_output_for_rect(rect);
             xdg_toplevel_set_fullscreen(surface->xdg_toplevel, output);
             surface->requested_output = output;
             pthread_mutex_unlock(&process_wayland.output_mutex);
@@ -314,14 +336,20 @@ static void wayland_surface_update_state_toplevel(struct wayland_surface *surfac
         }
         else
         {
-            xdg_toplevel_set_min_size(
-                surface->xdg_toplevel,
-                rect.right - rect.left,
-                rect.bottom - rect.top);
-            xdg_toplevel_set_max_size(
-                surface->xdg_toplevel,
-                rect.right - rect.left,
-                rect.bottom - rect.top);
+            int width = rect->right - rect->left, height = rect->bottom - rect->top;
+            xdg_toplevel_set_min_size(surface->xdg_toplevel, width, height);
+            xdg_toplevel_set_max_size(surface->xdg_toplevel, width, height);
+        }
+        if (process_wayland.xdg_toplevel_icon_manager_v1 && !surface->big_icon_buffer)
+        {
+            HICON big, small;
+            ICONINFO ii, ii_small;
+
+            big = get_window_icon(surface->hwnd, ICON_BIG, 0, &ii);
+            small = get_window_icon(surface->hwnd, ICON_SMALL, 0, &ii_small);
+
+            if (big) wayland_surface_set_icon(surface, ICON_BIG, &ii);
+            if (small) wayland_surface_set_icon(surface, ICON_SMALL, &ii_small);
         }
     }
     else
@@ -469,32 +497,6 @@ BOOL WAYLAND_WindowPosChanging(HWND hwnd, UINT swp_flags, BOOL shaped, const str
     return TRUE;
 }
 
-static HICON get_icon_info(HICON icon, ICONINFO *ii)
-{
-    return icon && NtUserGetIconInfo(icon, ii, NULL, NULL, NULL, 0) ? icon : NULL;
-}
-
-static HICON get_window_icon(HWND hwnd, UINT type, HICON icon, ICONINFO *ret)
-{
-    icon = get_icon_info(icon, ret);
-    if (!icon)
-    {
-        icon = get_icon_info((HICON)send_message(hwnd, WM_GETICON, type, 0), ret);
-        if (!icon)
-        {
-            UINT offset = (type == ICON_BIG) ? GCLP_HICON : GCLP_HICONSM;
-            icon = get_icon_info((HICON)NtUserGetClassLongPtrW(hwnd, offset), ret);
-        }
-        if (!icon && type == ICON_BIG)
-        {
-            icon = LoadImageW(0, (const WCHAR *)IDI_WINLOGO, IMAGE_ICON, 0, 0,
-                              LR_SHARED | LR_DEFAULTSIZE);
-            icon = get_icon_info(icon, ret);
-        }
-    }
-    return icon;
-}
-
 static int use_force_below_hack(void)
 {
     static int cached = -1;
@@ -521,7 +523,7 @@ void WAYLAND_WindowPosChanged(HWND hwnd, HWND insert_after, HWND owner_hint, UIN
     struct wayland_surface *toplevel_surface;
     struct wayland_client_surface *client;
     struct wayland_win_data *data, *toplevel_data;
-    BOOL managed, needs_icon;
+    BOOL managed;
 
     TRACE("hwnd %p new_rects %s after %p flags %08x\n", hwnd, debugstr_window_rects(new_rects), insert_after, swp_flags);
 
@@ -569,27 +571,7 @@ void WAYLAND_WindowPosChanged(HWND hwnd, HWND insert_after, HWND owner_hint, UIN
         wayland_win_data_update_wayland_state(data);
     }
 
-    needs_icon = data->wayland_surface && !data->wayland_surface->big_icon_buffer &&
-                 wayland_surface_is_toplevel(data->wayland_surface) &&
-                 process_wayland.xdg_toplevel_icon_manager_v1;
-
     wayland_win_data_release(data);
-
-    if (needs_icon)
-    {
-        HICON big, small;
-        ICONINFO ii, ii_small;
-        big = get_window_icon(hwnd, ICON_BIG, 0, &ii);
-        small = get_window_icon(hwnd, ICON_SMALL, 0, &ii_small);
-
-        if((data = wayland_win_data_get(hwnd)))
-        {
-            if (big) wayland_surface_set_icon(data->wayland_surface, ICON_BIG, &ii);
-            if (small) wayland_surface_set_icon(data->wayland_surface, ICON_SMALL, &ii_small);
-
-            wayland_win_data_release(data);
-        }
-    }
 }
 
 static void wayland_configure_window(HWND hwnd)
-- 
2.52.0

